### DISCUSSION TITLE
Is 3D Support still possible in the future?

### DISCUSSION THREAD

**Discussion by zach95564** [OndrejNepozitek/Edgar-Unity#91] [General]
Hey Onndrej, really appreciate all your hard work. Curious if you are still considering Edgar for 3D in the near future?

Cheers!

**Comment by OndrejNepozitek** [comment]
Hey! I'm currently working on a "pseudo-3D" version of Edgar. What I mean by that is that the core of the algorithm is still 2D but individual rooms can have different elevations. In practice, it means that no room can be placed on top of/below a different room.

In order to see the limitations and challenges of this version of the generator, I'm trying to replicate a level from Minecraft Dungeons - you can see an early WIP screenshot below. (The terrain outside of individual rooms was added in a post-processing script.)

![image](https://user-images.githubusercontent.com/2265359/102381595-b73aea00-3fc9-11eb-8685-0da0c0df1e2e.png)

There is one major limitation that will probably not get solved any time soon. The algorithm that computes how can two rooms be positioned so that they are properly connected works only on integer grids. That means that I must be able to describe an outline of a room as a collection of blocks/tiles. What can be done about that is that I can go through tiles that are only partially filled and say that the tile is completely filled instead and somehow trick the algorithm into thinking that the room template is tile-based. Unfortunately, the same thing cannot be done with doors. They must be aligned on the grid and their size must be a multiple of the grid unit size. Quite a lot of work would be needed to overcome this limitation.

Can you please describe your use case? How would you want to use Edgar in the 3D context?

Edit: One think I forgot to mention is that I currently plan to add the 3D support only to the PRO version of the asset.
